Linear Discriminant Analysis


In [1]:
import pandas as pd
import numpy as np
from sklearn.discriminant_analysis import LinearDiscriminantAnalysis
from sklearn.metrics import confusion_matrix
from sklearn.metrics import classification_report

In [2]:
data = pd.read_csv(r'C:\Users\SHREYA\Documents\heart.csv')
data.head()

age  sex  cp  trestbps  chol  fbs  restecg  thalach  exang  oldpeak  slope  \
0   63    1   3       145   233    1        0      150      0      2.3      0   
1   37    1   2       130   250    0        1      187      0      3.5      0   
2   41    0   1       130   204    0        0      172      0      1.4      2   
3   56    1   1       120   236    0        1      178      0      0.8      2   
4   57    0   0       120   354    0        1      163      1      0.6      2   

   ca  thal  target  
0   0     1       1  
1   0     2       1  
2   0     2       1  
3   0     2       1  
4   0     2       1

Attribute Information : 
1. age
2. sex
3. chest pain type (4 values)
4. resting blood pressure
5. serum cholestoral in mg/dl
6. fasting blood sugar > 120 mg/dl
7. resting electrocardiographic results (values 0,1,2)
8. maximum heart rate achieved
9. exercise induced angina
10. oldpeak = ST depression induced by exercise relative to rest
11. the slope of the peak exercise ST segment
12. number of major vessels (0-3) colored by flourosopy
13. thal: 0 = normal; 1 = fixed defect; 2 = reversable defect
14. target: 0= less chance of heart attack 1= more chance of heart attack


In [3]:
# spilting data in X and y as independent and dependent variable
X = data[['age','sex','cp','trestbps','chol','fbs','restecg','thalach','exang','oldpeak','slope','ca','thal']]
y = data[['target']]

In [4]:
lda = LinearDiscriminantAnalysis()
lda.fit(X, y)
yp= lda.predict(X)

C:\Users\SHREYA\anaconda3\lib\site-packages\sklearn\utils\validation.py:73: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  return f(**kwargs)


In [5]:
lda.decision_function(X)

array([ 2.19436212e+00,  1.25692319e+00,  3.60270662e+00,  2.80180710e+00,
        1.66586579e+00,  1.43367379e+00,  1.98791370e+00,  2.08113018e+00,
        1.91511101e+00,  2.94366182e+00,  8.93910210e-01,  4.45464423e+00,
        2.51828265e+00,  1.25156033e+00,  4.88462338e+00,  3.90795425e+00,
        5.28587056e+00,  1.97150542e+00,  9.09374039e-01,  3.03710936e+00,
       -3.20089805e-01,  2.68713871e+00,  2.05070155e+00,  3.54284244e-01,
        2.08314066e+00,  1.59565400e+00,  2.51899133e+00,  2.83481355e+00,
        2.81030026e+00,  1.27912382e+00,  4.19521891e+00,  2.32384639e-01,
        2.99542941e+00,  4.94081361e-01,  6.25728965e-01,  1.57713147e+00,
        5.26494527e+00,  1.14850538e+00,  3.87598667e+00,  3.18759378e+00,
        2.36369014e+00,  1.93801963e+00, -1.80352858e+00,  1.51460633e+00,
        3.35009060e+00,  2.71026653e+00,  3.51832326e+00,  2.86476438e+00,
        5.68555375e+00,  2.74749554e+00,  4.18037839e+00,  5.31587658e-02,
       -1.89386289e+00,  

In [6]:
confusion_matrix = confusion_matrix(y, yp)
print(confusion_matrix)

[[102  36]
 [ 12 153]]


In [7]:
print(classification_report(y, yp))

              precision    recall  f1-score   support

           0       0.89      0.74      0.81       138
           1       0.81      0.93      0.86       165

    accuracy                           0.84       303
   macro avg       0.85      0.83      0.84       303
weighted avg       0.85      0.84      0.84       303

